pkg/analysis/preferredmarkers/analyzer.go

```diff
@@ -0,0 +1,258 @@
+
+	return strings.Join(names, ", ")
+}
+
+// buildTextEdits generates the text edits to fix equivalent markers.
+// If preferredExists is true, all markers are deleted. Otherwise, the first
+// marker is replaced with the preferred identifier and the rest are deleted.
+func buildTextEdits(marks []markers.Marker, preferredIdentifier string, preferredExists bool) []analysis.TextEdit {
+	// Sort markers by position to ensure deterministic text edits
+	sort.Slice(marks, func(i, j int) bool {
+		return marks[i].Pos < marks[j].Pos
+	})
+
+	edits := make([]analysis.TextEdit, 0, len(marks))
+
+	// If the preferred marker doesn't exist, replace the first equivalent marker
+	if !preferredExists {
+		edits = append(edits, analysis.TextEdit{
+			Pos:     marks[0].Pos,
+			End:     marks[0].End,
+			NewText: []byte(buildReplacementMarker(marks[0], preferredIdentifier)),
+		})
+		marks = marks[1:] // Process remaining markers for deletion
+	}
+
+	// Delete remaining markers to avoid duplicates
+	// Note: We add 1 to the end position to include the newline character,
+	// which removes the entire line and prevents blank lines in the output.
+	// This works correctly for most cases. At end of file without a trailing
+	// newline, the go/analysis framework handles the extra position gracefully.
+	for _, mark := range marks {
+		edits = append(edits, analysis.TextEdit{
+			Pos:     mark.Pos,
+			End:     mark.End + 1, // +1 to include the newline character
+			NewText: []byte(""),
+		})
+	}
+
+	return edits
+}
+
+// reportMarkers generates a diagnostic report for markers that should be
+// replaced. This function handles the common logic for both field and type
+// reporting.
+func reportMarkers(pass *analysis.Pass, marks []markers.Marker, preferredIdentifier, elementName string, pos token.Pos, elementType string, preferredExists bool) {
+	if len(marks) == 0 {
+		return
+	}
+
+	markerWord := "marker"
+	if len(marks) > 1 {
+		markerWord = "markers"
+	}
+
+	message := fmt.Sprintf("%s %s uses %s %s, should use preferred marker %q instead",
+		elementType, elementName, markerWord, formatMarkerList(marks), preferredIdentifier)
+
+	fixMessage := "remove equivalent markers"
+	if !preferredExists {
+		fixMessage = fmt.Sprintf("replace with %q", preferredIdentifier)
+	}
+
+	pass.Report(analysis.Diagnostic{
+		Pos:     pos,
+		Message: message,
+		SuggestedFixes: []analysis.SuggestedFix{
+			{
+				Message:   fixMessage,
+				TextEdits: buildTextEdits(marks, preferredIdentifier, preferredExists),
+			},
+		},
+	})
+}
+
+// formatExpressions formats marker expressions as a string suffix.
+// Returns ":=value" for unnamed expressions or ":key1=val1,key2=val2" for named expressions.
+func formatExpressions(expressions map[string]string) string {
+	if len(expressions) == 0 {
+		return ""
+	}
+
+	// Check for unnamed expression (simple case like "+marker:=value")
+	if unnamedValue, ok := expressions[markers.UnnamedExpression]; ok {
+		return ":=" + unnamedValue
```

Comment:
I don't think `:=` is universally compatible, just use `=`.

Kubebuilder supports `:=`, but to my knowledge, the `+k8s` markers do not